styx-scheduler-service/src/test/java/com/spotify/styx/state/QueuedStateManagerTest.java

```diff
@@ -616,4 +589,125 @@ public void shouldThrowRuntimeException() throws Exception {
+  }
+
+  @Test
+  public void shouldNotUpdateResourceCountersOnSuccess() throws Exception {
+    givenState(INSTANCE, State.TERMINATED);
+    receiveEvent(Event.success(INSTANCE));
+    verify(transaction, never()).updateCounter(eq(shardedCounter), anyString(), anyInt());
+  }
+
+  @Test
+  public void shouldUpdateResourceCountersOnHaltWhileRunning() throws Exception {
+    givenState(INSTANCE, State.RUNNING);
+    receiveEvent(Event.halt(INSTANCE));
+    verify(transaction).updateCounter(shardedCounter, "resource1", -1);
+  }
+
+  @Test
+  public void shouldUpdateResourceCountersOnTimeoutWhileRunning() throws Exception {
+    givenState(INSTANCE, State.RUNNING);
+    receiveEvent(Event.timeout(INSTANCE));
+    verify(transaction).updateCounter(shardedCounter, "resource1", -1);
+  }
+
+  @Test
+  public void shouldNotUpdateResourceCountersOnTimeoutWhileQueued() throws Exception {
+    givenState(INSTANCE, State.QUEUED);
+    receiveEvent(Event.timeout(INSTANCE));
+    verify(transaction, never()).updateCounter(eq(shardedCounter), anyString(), anyInt());
+  }
+
+  @Test
+  public void shouldNotUpdateResourceCountersOnStopAfterFailed() throws Exception {
+    givenState(INSTANCE, State.FAILED);
+    receiveEvent(Event.stop(INSTANCE));
+    verify(transaction, never()).updateCounter(eq(shardedCounter), anyString(), anyInt());
+  }
+
+  @Test
+  public void shouldNotUpdateResourceCountersOnRunErrorWhileQueued() throws Exception {
+    givenState(INSTANCE, State.QUEUED);
+    receiveEvent(Event.runError(INSTANCE, "random error"));
+    verify(transaction, never()).updateCounter(eq(shardedCounter), anyString(), anyInt());
+  }
+
+  @Test
+  public void shouldUpdateResourceCountersOnRunErrorWhileRunning() throws Exception {
+    givenState(INSTANCE, State.RUNNING);
+    receiveEvent(Event.runError(INSTANCE, "random error"));
+    verify(transaction).updateCounter(shardedCounter, "resource1", -1);
+  }
+
+  @Test
+  public void shouldNotUpdateResourceCountersOnRetryAfterFailed() throws Exception {
+    givenState(INSTANCE, State.FAILED);
+    receiveEvent(Event.retryAfter(INSTANCE,10));
+    verify(transaction, never()).updateCounter(eq(shardedCounter), anyString(), anyInt());
+  }
+
+  @Test
+  public void shouldNotUpdateResourceCountersOnRetryAfterTerminated() throws Exception {
+    givenState(INSTANCE, State.TERMINATED);
+    receiveEvent(Event.retryAfter(INSTANCE, 10));
+    verify(transaction, never()).updateCounter(eq(shardedCounter), anyString(), anyInt());
+  }
+
+
+  public void givenState(WorkflowInstance instance, State state) throws IOException {
+    when(transaction.activeState(instance))
+        .thenReturn(Optional.of(PersistentWorkflowInstanceState.builder()
+                                    .counter(17)
+                                    .timestamp(NOW.minusMillis(1))
+                                    .state(state)
+                                    .data(STATE_DATA_1)
```

Comment:
This will sometimes set resourceIds in a state where they're normally not set (QUEUED)... Not sure if it's a problem.

Reply:
It shouldn't be a problem. The reason I've done it, is because if I set data to `StateData.zero()` that will make resourceids be empty, which will fail the `.isPresent()` check and won't make the `.updateCounter` call. This is important for the negative tests: I'm making sure the resource ids are there, and the only reason `.updateCounter` won't be called is due to the type of transition that is taking place (i.e. the condition boils down to `isConsuming` and `!isConsuming` check and the event type for the `dequeue` case) .